src/cargo/core/compiler/mod.rs

```diff
@@ -713,20 +759,22 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
     let build_script_outputs = Arc::clone(&cx.build_script_outputs);
     let package_id = unit.pkg.package_id();
     let manifest_path = PathBuf::from(unit.pkg.manifest_path());
-    let relative_manifest_path = manifest_path
-        .strip_prefix(cx.bcx.ws.root())
-        .unwrap_or(&manifest_path)
-        .to_owned();
     let target = Target::clone(&unit.target);
     let mut output_options = OutputOptions::new(cx, unit);
     let script_metadata = cx.find_build_script_metadata(unit);
+
     let failed_scrape_units = Arc::clone(&cx.failed_scrape_units);
-    let hide_diagnostics_for_scrape_unit = unit.mode.is_doc_scrape()
-        && unit.target.doc_scrape_examples().is_unset()
+    let hide_diagnostics_for_scrape_unit = cx.bcx.unit_can_fail_for_docscraping(unit)
         && !matches!(cx.bcx.config.shell().verbosity(), Verbosity::Verbose);
+    let failed_scrape_diagnostic = make_failed_scrape_diagnostic(
```

Comment:
Yeah I originally tried to avoid this allocation. For example, I did something like:

```rust
fn make_failed_scrape_diagnostic(...) -> impl FnOnce() -> String {
  /* context .. */
  || format!("...")
}
```

However because this closure gets moved into `Work::new` it has to be `Send` and `Sync`, and a lot of the environment  of the closure wasn't sendable. Eg once I change `top_line` to use `format_args`, then I get:

```
error[E0277]: `core::fmt::Opaque` cannot be shared between threads safely
   --> src/cargo/core/compiler/mod.rs:309:25
    |
309 |       return Ok(Work::new(move |state| {
    |  _______________---------_^
    | |               |
    | |               required by a bound introduced by this call
310 | |         // Artifacts are in a different location than typical units,
311 | |         // hence we must assure the crate- and target-dependent
312 | |         // directory is present.
...   |
451 | |         Ok(())
452 | |     }));
    | |_____^ `core::fmt::Opaque` cannot be shared between threads safely
note: required because it's used within this closure
   --> src/cargo/core/compiler/mod.rs:223:5
    |
223 |     move || format!(
    |     ^^^^^^^
note: required because it appears within the type `impl FnOnce() -> std::string::String`
   --> src/cargo/core/compiler/mod.rs:217:6
    |
217 | ) -> impl FnOnce() -> String {
```